# Scaling

**Note:** You will need to search many new functions in NumPy documentation for many of these tasks. These are the kind of open-ended questions that a data scientist needs to tackle. This exercise will be a good learning experience. 

## 1. Creating a dataset

Let's create a hypothetical dataset with data about athletes, their physical attributes, and their sports records. 

### 1.1 
Create an array to identify each athlete uniquely. The array should have all unique five-digit integers. The shape of the array must be one thousand elements rows. (1000)

In [1]:
# importing necessary libraries 
import numpy as np
# Seting a seed so that the random command generates similar numbers for everyone.  
np.random.seed(0) 
# Suppressing the scientific notation so you can read numbers in the output. 
np.set_printoptions(suppress=True)

In [2]:
athlete_id = np.random.choice(np.random.randint(10000, 99999, size= (5000)),size = (1000), replace=False )

In [3]:
athlete_id.shape

(1000,)

In [4]:
athlete_id

array([79005, 95341, 34443, 78826, 77966, 57141, 46639, 26975, 77029,
       54284, 70241, 69005, 18232, 54928, 22877, 40145, 22754, 84072,
       75112, 42102, 86965, 16034, 66137, 16463, 87528, 42723, 65721,
       85378, 10739, 31273, 81803, 34262, 67185, 46560, 63721, 21391,
       61235, 93575, 60972, 82964, 52810, 47028, 83041, 96085, 46884,
       37398, 50452, 23829, 97391, 51624, 88573, 49351, 66062, 18211,
       61566, 11221, 33306, 25409, 74309, 43843, 43502, 41388, 37181,
       60533, 52098, 94043, 21071, 64342, 36575, 63525, 62156, 16845,
       51852, 13418, 38341, 42329, 10523, 47551, 76884, 79713, 92999,
       11107, 84702, 86076, 15241, 12987, 39987, 32302, 55553, 77792,
       78828, 74923, 81562, 33704, 11339, 85802, 44074, 24445, 40504,
       15479, 86473, 14010, 69530, 46076, 61629, 39522, 67088, 64921,
       27729, 63076, 41289, 32579, 25393, 94466, 53986, 87931, 45571,
       42371, 57456, 82709, 88878, 36703, 14845, 57170, 11064, 43671,
       37821, 68085,

### 1.2 
Check if all the entries in the "athlete_id" column are unique.

**Hint:** Find the number of unique entries in the athlete_id array and compare that to the total number of elements in the array. 

In [5]:
np.unique(athlete_id).shape

(993,)

Seven values repeat twice. You will have to remove duplicates and add three unique numbers. 

### 1.3 
Find the duplicate values and remove them. 

In [6]:
athlete_list = athlete_id.tolist()

In [7]:
athlete_list

[79005,
 95341,
 34443,
 78826,
 77966,
 57141,
 46639,
 26975,
 77029,
 54284,
 70241,
 69005,
 18232,
 54928,
 22877,
 40145,
 22754,
 84072,
 75112,
 42102,
 86965,
 16034,
 66137,
 16463,
 87528,
 42723,
 65721,
 85378,
 10739,
 31273,
 81803,
 34262,
 67185,
 46560,
 63721,
 21391,
 61235,
 93575,
 60972,
 82964,
 52810,
 47028,
 83041,
 96085,
 46884,
 37398,
 50452,
 23829,
 97391,
 51624,
 88573,
 49351,
 66062,
 18211,
 61566,
 11221,
 33306,
 25409,
 74309,
 43843,
 43502,
 41388,
 37181,
 60533,
 52098,
 94043,
 21071,
 64342,
 36575,
 63525,
 62156,
 16845,
 51852,
 13418,
 38341,
 42329,
 10523,
 47551,
 76884,
 79713,
 92999,
 11107,
 84702,
 86076,
 15241,
 12987,
 39987,
 32302,
 55553,
 77792,
 78828,
 74923,
 81562,
 33704,
 11339,
 85802,
 44074,
 24445,
 40504,
 15479,
 86473,
 14010,
 69530,
 46076,
 61629,
 39522,
 67088,
 64921,
 27729,
 63076,
 41289,
 32579,
 25393,
 94466,
 53986,
 87931,
 45571,
 42371,
 57456,
 82709,
 88878,
 36703,
 14845,
 57170,
 11064,


In [8]:
athlete_set = set(athlete_list)

In [9]:
athlete_set

{10131,
 10215,
 10383,
 10523,
 10739,
 10846,
 10851,
 10922,
 10972,
 11009,
 11064,
 11107,
 11184,
 11221,
 11339,
 11612,
 11677,
 11740,
 11924,
 11945,
 12042,
 12121,
 12195,
 12205,
 12240,
 12352,
 12722,
 12987,
 13041,
 13107,
 13134,
 13244,
 13278,
 13405,
 13418,
 13513,
 13627,
 13764,
 13846,
 13938,
 14010,
 14078,
 14322,
 14409,
 14562,
 14641,
 14845,
 14854,
 14893,
 14930,
 14961,
 14973,
 15037,
 15162,
 15241,
 15383,
 15393,
 15479,
 15791,
 15850,
 15912,
 16034,
 16076,
 16238,
 16421,
 16448,
 16462,
 16463,
 16521,
 16845,
 16949,
 17213,
 17299,
 17426,
 17509,
 17524,
 17661,
 17997,
 18197,
 18211,
 18232,
 18266,
 18489,
 18803,
 18819,
 18824,
 18947,
 18989,
 19015,
 19042,
 19089,
 19097,
 19191,
 19265,
 19307,
 19350,
 19359,
 19630,
 19654,
 19740,
 19748,
 19820,
 19986,
 20088,
 20417,
 20525,
 20643,
 20670,
 20709,
 20738,
 20751,
 20787,
 21063,
 21071,
 21178,
 21351,
 21390,
 21391,
 21498,
 21636,
 21723,
 22013,
 22079,
 22127,
 22180,


In [10]:
len(athlete_set)

993

Since the set is now sorted, you can identify the ids which are not present in the set. For instance, you can add 10001, 10002, 10003, etc., which are not present in the data. 

In [11]:
for number in [10001, 10002, 10003,10004,10005,10006,10007]:
    athlete_set.add(number)

In [12]:
len(athlete_set)

1000

In [13]:
athlete_set

{86016,
 53252,
 51205,
 43016,
 61454,
 24591,
 49168,
 30737,
 90134,
 22552,
 71704,
 98330,
 65564,
 53277,
 28701,
 63525,
 16421,
 88105,
 61483,
 20525,
 73774,
 73776,
 57394,
 22579,
 90163,
 49206,
 77880,
 18489,
 86076,
 43069,
 16448,
 12352,
 84034,
 86084,
 61510,
 49225,
 14409,
 26700,
 49228,
 16462,
 16463,
 26712,
 79963,
 65627,
 51294,
 39011,
 84072,
 43112,
 47212,
 57456,
 71793,
 59507,
 79988,
 39030,
 67704,
 96377,
 75901,
 61566,
 24705,
 24707,
 75911,
 16521,
 47241,
 86155,
 77966,
 63630,
 88206,
 10383,
 32918,
 80027,
 24731,
 96416,
 20643,
 75940,
 45220,
 37045,
 37047,
 65721,
 73914,
 61629,
 32957,
 20670,
 90304,
 92354,
 92357,
 49351,
 39111,
 26825,
 75978,
 98507,
 46444,
 82120,
 75983,
 71888,
 30929,
 82130,
 55510,
 22747,
 57565,
 55518,
 69855,
 30943,
 73953,
 22754,
 14562,
 43236,
 20709,
 67812,
 94432,
 69864,
 63721,
 94442,
 55531,
 80105,
 84205,
 92398,
 22772,
 94458,
 78074,
 43260,
 37117,
 26878,
 55553,
 94466,
 28931,


In [14]:
athlete_list = list(athlete_set)

In [15]:
athlete_list.sort()

In [16]:
athlete_list

[10001,
 10002,
 10003,
 10004,
 10005,
 10006,
 10007,
 10131,
 10215,
 10383,
 10523,
 10739,
 10846,
 10851,
 10922,
 10972,
 11009,
 11064,
 11107,
 11184,
 11221,
 11339,
 11612,
 11677,
 11740,
 11924,
 11945,
 12042,
 12121,
 12195,
 12205,
 12240,
 12352,
 12722,
 12987,
 13041,
 13107,
 13134,
 13244,
 13278,
 13405,
 13418,
 13513,
 13627,
 13764,
 13846,
 13938,
 14010,
 14078,
 14322,
 14409,
 14562,
 14641,
 14845,
 14854,
 14893,
 14930,
 14961,
 14973,
 15037,
 15162,
 15241,
 15383,
 15393,
 15479,
 15791,
 15850,
 15912,
 16034,
 16076,
 16238,
 16421,
 16448,
 16462,
 16463,
 16521,
 16845,
 16949,
 17213,
 17299,
 17426,
 17509,
 17524,
 17661,
 17997,
 18197,
 18211,
 18232,
 18266,
 18489,
 18803,
 18819,
 18824,
 18947,
 18989,
 19015,
 19042,
 19089,
 19097,
 19191,
 19265,
 19307,
 19350,
 19359,
 19630,
 19654,
 19740,
 19748,
 19820,
 19986,
 20088,
 20417,
 20525,
 20643,
 20670,
 20709,
 20738,
 20751,
 20787,
 21063,
 21071,
 21178,
 21351,
 21390,
 21391,


In [17]:
athlete_id = np.array(athlete_list)

In [18]:
athlete_id.shape

(1000,)

Now we have an array of 1000 unique numbers in a column format. 

So in the dataset, we have the first column set.

### 1.4 
Create a column for the height of the athletes. 

The height column must follow a few rules. 

1. Height of an athlete is a natural variable. So it would be best if you used normal distribution to create the height data. 
2. You also need to fix the maximum and minimum heights.
3. Height need not be unique. 
4. Height needs to be in whole centimeters (integers)

Assumptions - 
1. Shortest person's height - 160 cm 
2. Tallest person's height - 240 cm
3. Average person's height - 200 cm

In [19]:
height_array = np.random.normal(200, 20, size=(1000))

In [20]:
height_array = np.round_(height_array, decimals=0)

In [21]:
height_array = height_array.astype(int)

In [22]:
height_array.shape

(1000,)

In [23]:
height_array

array([180, 229, 181, 172, 230, 184, 192, 218, 205, 179, 255, 190, 195,
       175, 188, 221, 174, 197, 203, 222, 179, 188, 191, 200, 241, 208,
       194, 188, 192, 231, 217, 205, 197, 170, 196, 198, 200, 203, 201,
       219, 187, 214, 153, 177, 186, 205, 213, 197, 214, 253, 192, 184,
       182, 211, 122, 203, 197, 173, 222, 199, 208, 196, 198, 184, 168,
       221, 153, 201, 208, 211, 184, 216, 227, 203, 193, 215, 188, 199,
       203, 182, 201, 228, 226, 216, 200, 226, 186, 192, 220, 215, 181,
       167, 213, 206, 216, 205, 207, 211, 170, 210, 196, 190, 208, 219,
       232, 192, 221, 212, 198, 215, 183, 230, 194, 253, 234, 229, 181,
       181, 204, 179, 181, 189, 235, 187, 217, 208, 178, 191, 175, 193,
       198, 207, 171, 205, 201, 218, 219, 204, 194, 194, 169, 183, 212,
       192, 210, 224, 233, 190, 205, 221, 219, 219, 204, 190, 209, 239,
       178, 180, 222, 207, 218, 184, 155, 200, 200, 232, 201, 213, 190,
       192, 198, 184, 172, 219, 218, 186, 200, 208, 218, 203, 20

The problem with normal distribution is that there will always be outliers. For instance, in the list above, you can find values like 138, 272 which are undesirable. Although in real life, you will find people who are exceptionally tall or short. But in this case, we want to consider the people with heights between 160 and 240. 

So you will have to replace those values with other values from the acceptable range. 

Now, the task is to ensure that all the randomly picked values are nomally distributed and fall within an acceptable range. 

Let's think of ways of meeting all the said conditions. 

1. Create a larger sample of normally distributed integers than needed
2. Use the filter command to extract all the numbers between the decided range. 
3. Finally, you need to ensure the number of observations created to the required number. 

There is a flow with this method. In the end, you might not end up with a normally distributed range. So it might not work. 

As an alternative, there is a more advanced library called scipy, it builds on top of NumPy and adds a lot more functionality. The scipy library has a function called **truncnorm**

Reading material for the function:

1. [Official documentation for truncnorm](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.truncnorm.html)

2. [Explanation for the usage of truncnorm](https://stackoverflow.com/questions/18441779/how-to-specify-upper-and-lower-limits-when-using-numpy-random-normal)

Since the scipy library is not covered, the following code block will create the necessary observations. 

In [24]:
import scipy.stats

lower = 160
upper = 240
mu = 200
sigma = 20
N = 1000

samples = scipy.stats.truncnorm.rvs(
          (lower-mu)/sigma,(upper-mu)/sigma,loc=mu,scale=sigma,size=N)

In [25]:
height_array = np.round_(samples, decimals=0)
height_array = height_array.astype(int)

In [26]:
height_array

array([225, 196, 218, 216, 219, 170, 219, 198, 197, 212, 211, 205, 204,
       190, 213, 210, 187, 209, 208, 181, 213, 211, 215, 214, 184, 203,
       192, 212, 195, 196, 208, 223, 179, 175, 238, 199, 199, 219, 235,
       188, 192, 209, 220, 188, 164, 214, 200, 180, 182, 200, 177, 215,
       197, 195, 175, 191, 202, 184, 189, 189, 206, 182, 190, 218, 183,
       189, 220, 187, 199, 189, 193, 223, 202, 209, 231, 217, 181, 175,
       209, 210, 223, 220, 196, 170, 231, 178, 219, 202, 181, 215, 208,
       223, 211, 237, 201, 239, 184, 237, 210, 216, 196, 171, 189, 180,
       204, 179, 164, 220, 224, 194, 211, 230, 213, 201, 170, 191, 182,
       195, 209, 190, 198, 199, 209, 186, 202, 188, 207, 198, 177, 228,
       219, 196, 217, 186, 227, 200, 186, 192, 198, 197, 222, 209, 207,
       219, 214, 197, 175, 212, 170, 180, 189, 188, 179, 178, 207, 190,
       197, 199, 224, 193, 198, 209, 209, 229, 201, 171, 171, 212, 212,
       170, 179, 180, 174, 219, 237, 177, 177, 203, 218, 181, 19

### 1.5 
Create a length of the jump array. 

The length of an athlete's jump is a variable that depends on a lot of different factors, like the physical strength of the athlete, training, their natural talent and so on. So it will be fair to draw variables within a limit randomly. The length is measured in meters, with the least count 1 cm. Create the length of the jump parameter for the data. 

The length of the average long jump is 6 to 7 meters. The world record is 8.95 m. Hence, let's assume the highest long jump that we recorded in our data set to be 8.00m and the lowest limit to be 5.00m. 

To generate random numbers between 5 and 8. with uniform distribution. 
Why uniform distribution? Again the same most values will be close to the mean; rarely people will be good or bad. 

In [27]:
jump_array = np.random.uniform(low=5.00, high=8.00, size=(1000))

In [28]:
jump_array = np.round_(jump_array, decimals=2)

In [29]:
jump_array

array([5.79, 7.31, 6.01, 7.25, 5.45, 6.26, 5.08, 6.11, 5.05, 7.38, 5.81,
       7.34, 7.39, 5.24, 6.55, 5.34, 7.93, 7.21, 5.36, 6.53, 7.97, 7.44,
       7.19, 7.63, 7.88, 6.94, 6.49, 5.88, 5.67, 5.67, 6.35, 6.3 , 7.7 ,
       6.94, 5.16, 5.17, 6.9 , 5.14, 7.75, 5.26, 5.17, 7.19, 6.41, 5.56,
       6.05, 6.78, 5.17, 7.39, 5.05, 6.32, 6.94, 7.64, 6.27, 6.26, 6.25,
       7.85, 6.78, 5.87, 6.14, 6.06, 5.57, 6.54, 6.6 , 5.36, 6.14, 6.79,
       6.28, 7.48, 7.95, 5.74, 5.66, 5.87, 7.59, 7.82, 6.3 , 7.53, 6.01,
       7.66, 5.91, 6.03, 7.57, 6.23, 7.35, 7.87, 5.14, 7.04, 5.68, 5.32,
       6.03, 6.27, 5.75, 5.99, 5.86, 6.12, 5.71, 6.53, 5.81, 6.83, 5.78,
       7.22, 6.71, 7.6 , 7.87, 7.04, 5.05, 6.68, 7.02, 6.64, 7.76, 5.19,
       6.22, 6.75, 6.62, 6.54, 6.23, 6.29, 6.77, 6.48, 7.35, 5.85, 5.09,
       7.62, 7.59, 5.47, 5.12, 6.21, 6.58, 5.59, 6.32, 5.54, 6.73, 6.38,
       5.27, 5.77, 7.2 , 7.47, 5.83, 6.46, 5.38, 7.41, 5.44, 6.92, 6.94,
       5.74, 6.31, 5.24, 6.24, 5.57, 7.94, 6.66, 5.

In [30]:
jump_array.shape

(1000,)

So now we have three arrays, unique athlete ID, normally distributed height array and uniformly distributed jump vector. 

### 1.6 
Combine the three arrays to create a dataset. 

The form of the dataset should be 

"Athlete_id", "Height", "Long_jump" and total 1000 reading of this form. 

hint: 
Take the sequences of 1-D arrays and stack them as columns to make a single 2-D array. 

In [31]:
data = np.column_stack((athlete_list,height_array,jump_array))

In [32]:
data

array([[10001.  ,   225.  ,     5.79],
       [10002.  ,   196.  ,     7.31],
       [10003.  ,   218.  ,     6.01],
       ...,
       [99733.  ,   195.  ,     7.09],
       [99874.  ,   204.  ,     5.84],
       [99974.  ,   204.  ,     5.39]])

Now you have created a dataset in the form that you need. 

## 2 Scaling Data

The next task is to perform a unit conversion so that the height (currently in cms) and length of jump (currently in m) are in the same unit. 


### 2.1 
Let's convert the height of the students to meters. 

Divide all values in the middle column by 100. 

#### Concept of vector division 

assume you have a vector 

a = 4,6,12

You want to divide the first element by 2, the second by three and the third by 6. 

#### Example 1

In [33]:
a = np.array([4,6,12])

In [34]:
b = np.array([2,3,6])

In [35]:
np.divide(a,b)

array([2., 2., 2.])

#### Example 2

In [36]:
a = np.array([[4,6,12], [6, 9, 18]])

In [37]:
a

array([[ 4,  6, 12],
       [ 6,  9, 18]])

In [38]:
np.divide(a,b)

array([[2., 2., 2.],
       [3., 3., 3.]])

You can use the divide function to perform row/column-wise operations. There are rules to follow while ensuring that such procedures do not through errors. You can read about thse rules [here](https://www.geeksforgeeks.org/python-broadcasting-with-numpy-arrays/#:~:text=Broadcasting%20Rules%3A&text=The%20two%20arrays%20are%20compatible,are%20compatible%20with%20all%20dimensions.)

This is called dimensional compatibility. You will learn more about this later in the program. 

In [39]:
modifier = np.array([1,100,1])

In [40]:
data = np.divide(data,modifier)

In [41]:
data

array([[10001.  ,     2.25,     5.79],
       [10002.  ,     1.96,     7.31],
       [10003.  ,     2.18,     6.01],
       ...,
       [99733.  ,     1.95,     7.09],
       [99874.  ,     2.04,     5.84],
       [99974.  ,     2.04,     5.39]])

### 2.2 
Calculate the length of the jump to height ratio. Higher this ratio better will be the student at the long jump. 

You will have to slice the array to extract the required columns from the matrix, perform the operation, and ultimately add the new column to the original matrix. 

In [42]:
product = data [0:, 2] / data[0:,1] 

In [43]:
product = np.round_(product, decimals=2)

In [44]:
data = np.column_stack((data,product))

In [45]:
data

array([[10001.  ,     2.25,     5.79,     2.57],
       [10002.  ,     1.96,     7.31,     3.73],
       [10003.  ,     2.18,     6.01,     2.76],
       ...,
       [99733.  ,     1.95,     7.09,     3.64],
       [99874.  ,     2.04,     5.84,     2.86],
       [99974.  ,     2.04,     5.39,     2.64]])

### 2.3 
Find the candidate ID with the highest length to height ratio. 

In [46]:
product.max()

4.95

In [47]:
np.where(product == 4.9)

(array([], dtype=int64),)

In [48]:
data[374,:]

array([43552.  ,     1.62,     6.5 ,     4.01])

The athlete with ID '43552' is the most efficient long jumper 

## Conclusion 

1. You learnt the use of official documentation and tools like stack overflow to reduce your effort in writing code. 
2. You used NumPy and scipy to create a dataset and perform operations on it 
3. You performed EDA found out the most efficient jumper. 